services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -127,16 +121,17 @@ private void doDownloadDirectory(CompletableFuture<CompletedDirectoryDownload> r
                                                            .build();
 
         Collection<FailedFileDownload> failedFileDownloads = new ConcurrentLinkedQueue<>();
-        List<CompletableFuture<?>> futures = new ArrayList<>();
+        List<CompletableFuture<CompletedFileDownload>> futures = new ArrayList<>();
 
-        listObjectsRecursivelyHelper.s3Objects(request).subscribe(s3Object -> {
+        listObjectsHelper.listS3ObjectsRecursively(request).subscribe(s3Object -> {
             log.debug(() -> "s3Object key: " + s3Object.key());
             futures.add(downloadSingleFile(downloadDirectoryRequest,
                                            failedFileDownloads,
                                            s3Object));
         }).whenComplete((r, t) -> {
             if (t != null) {
-                returnFuture.completeExceptionally(SdkClientException.create("Failed to call ListObjectsV2  ", t));
+                returnFuture.completeExceptionally(SdkClientException.create("Failed to call ListObjectsV2", t));
+                cleanUpDirectory(downloadDirectoryRequest.destinationDirectory());
```

Comment:
We should probably only call this if the SDK was responsible for creating the directory. If the directory already existed, I don't think we should delete it.

Reply:
To close the loop, discussed offline and agreed that we should move the logic to create directories in FileAsyncResponseTransformer and FileResponseTransformer. https://github.com/aws/aws-sdk-java-v2/pull/3018/